bool reset
int32 goal_idx
bool use_predefined